File: app/src/main/java/com/pavelprymak/propodcast/presentation/viewModels/PodcastInfoViewModel.java
package com.pavelprymak.propodcast.presentation.viewModels;

import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.pavelprymak.propodcast.model.network.PodcastApiController;
import com.pavelprymak.propodcast.model.network.pojo.podcastById.EpisodesItem;
import com.pavelprymak.propodcast.model.network.pojo.podcastById.PodcastResponse;
import com.pavelprymak.propodcast.model.network.pojo.podcasts.PodcastItem;
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoImpl;
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx;
import com.pavelprymak.propodcast.presentation.common.StatesBatch;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class PodcastInfoViewModel extends ViewModel {
    private final StatesBatch<PodcastResponse> mPodcastDataBatch = new StatesBatch<>();
    private final MutableLiveData<List<PodcastItem>> mRecommendData = new MutableLiveData<>();
    private final PodcastRepoRx mRepo = new PodcastRepoImpl(PodcastApiController.getInstance().getPodcastApi());
    private final String mId;

    PodcastInfoViewModel(String mId) {
        this.mId = mId;
    }

    public void preparePodcastInfoData() {
        if (mPodcastDataBatch.getData().getValue() == null) {
            mPodcastDataBatch.postLoading(true);
            Disposable disposable = mRepo.getPodcastById(mId, 0)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(mPodcastDataBatch::postData, mPodcastDataBatch::postError);
        }
    }

    public void loadMoreEpisoded() {
        PodcastResponse currentPodcastResponse = mPodcastDataBatch.getData().getValue();
        if (currentPodcastResponse != null && currentPodcastResponse.getEpisodes() != null && currentPodcastResponse.getEpisodes().size() < currentPodcastResponse.getTotalEpisodes()) {
            mPodcastDataBatch.postLoading(true);
            Disposable disposable = mRepo.getPodcastById(mId, currentPodcastResponse.getNextEpisodePubDate())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(podcastResponse -> {
                        if (podcastResponse != null) {
                            //concat episodes lists (current and new)
                            List<EpisodesItem> currentEpisodesList = currentPodcastResponse.getEpisodes();
                            if (podcastResponse.getEpisodes() != null) {
                                currentEpisodesList.addAll(podcastResponse.getEpisodes());
                            }
                            podcastResponse.setEpisodes(currentEpisodesList);
                            mPodcastDataBatch.postData(podcastResponse);
                        }
                    }, mPodcastDataBatch::postError);
        }

    }

    public StatesBatch<PodcastResponse> getPodcastDataBatch() {
        return mPodcastDataBatch;
    }

    public void removePodcastDataBatchObservers(LifecycleOwner lifecycleOwner) {
        mPodcastDataBatch.getData().removeObservers(lifecycleOwner);
        mPodcastDataBatch.getLoading().removeObservers(lifecycleOwner);
        mPodcastDataBatch.getError().removeObservers(lifecycleOwner);
    }

    public LiveData<List<PodcastItem>> getRecommendData() {
        if (mRecommendData.getValue() == null || mRecommendData.getValue().size() == 0) {
            Disposable disposable = mRepo.getPodcastRecommendations(mId)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(recommendations -> mRecommendData.setValue(recommendations.getRecommendations()), error -> {
                    });
        }
        return mRecommendData;
    }

}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/viewModels/PodcastInfoViewModel.kt
package com.pavelprymak.propodcast.presentation.viewModels

import androidx.lifecycle.LifecycleOwner
import com.pavelprymak.propodcast.model.network.pojo.podcastById.EpisodesItem
import com.pavelprymak.propodcast.model.network.pojo.podcastById.PodcastResponse
import com.pavelprymak.propodcast.model.network.pojo.podcasts.PodcastItem
import com.pavelprymak.propodcast.model.network.pojo.recommendations.RecommendationsResponse
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx
import com.pavelprymak.propodcast.presentation.common.SimpleViewModel
import com.pavelprymak.propodcast.presentation.common.StatesBatch
import com.pavelprymak.propodcast.presentation.common.bindToStatesBatch
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class PodcastInfoViewModel(private val mRepo: PodcastRepoRx) : SimpleViewModel() {
    private val mPodcastDataBatch = StatesBatch<PodcastResponse>()
    private val mRecommendDataBatch = StatesBatch<List<PodcastItem>?>()
    private var mId = ""

    fun setItemId(mId: String) {
        this.mId = mId
    }

    fun prepareRecommendData(): StatesBatch<List<PodcastItem>?> {
        if (mId.isEmpty()) throw IllegalArgumentException()
        if (mRecommendDataBatch.data.value == null) {
            mRepo.getPodcastRecommendations(mId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map { recommendResponse: RecommendationsResponse -> recommendResponse.recommendations }
                .bindToStatesBatch(mRecommendDataBatch)
                .untilDestroy()
        }
        return mRecommendDataBatch
    }

    fun preparePodcastInfoData(): StatesBatch<PodcastResponse> {
        if (mId.isEmpty()) throw IllegalArgumentException()
        if (mPodcastDataBatch.data.value == null) {
            mPodcastDataBatch.postLoading(true)
            mRepo.getPodcastById(mId, 0)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .bindToStatesBatch(mPodcastDataBatch)
                .untilDestroy()
        }
        return mPodcastDataBatch
    }

    fun loadMoreEpisodes() {
        if (mId.isEmpty()) throw IllegalArgumentException()
        mPodcastDataBatch.data.value?.let { prevPodcastResponse ->
            prevPodcastResponse.episodes?.let { prevEpisodes ->
                if (prevEpisodes.size < prevPodcastResponse.totalEpisodes) {
                    mPodcastDataBatch.postLoading(true)
                    mRepo.getPodcastById(mId, prevPodcastResponse.nextEpisodePubDate)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        //concat episodes lists (current and new)
                        .map { podcastInfo: PodcastResponse ->
                            concatPrevWithCurrentEpisodes(
                                podcastInfo,
                                prevEpisodes
                            )
                        }
                        .bindToStatesBatch(mPodcastDataBatch)
                        .untilDestroy()
                }
            }
        }
    }

    private fun concatPrevWithCurrentEpisodes(
        podcastInfo: PodcastResponse,
        episodes: List<EpisodesItem>
    ): PodcastResponse {
        val prevEpisodesList: ArrayList<EpisodesItem> = episodes as ArrayList<EpisodesItem>
        podcastInfo.episodes?.let { currentEpisodes ->
            prevEpisodesList.addAll(currentEpisodes)
        }
        podcastInfo.episodes = prevEpisodesList
        return podcastInfo
    }

    fun removePodcastDataBatchObservers(lifecycleOwner: LifecycleOwner) {
        mPodcastDataBatch.data.removeObservers(lifecycleOwner)
        mPodcastDataBatch.loading.removeObservers(lifecycleOwner)
        mPodcastDataBatch.error.removeObservers(lifecycleOwner)
        mRecommendDataBatch.data.removeObservers(lifecycleOwner)
        mRecommendDataBatch.loading.removeObservers(lifecycleOwner)
        mRecommendDataBatch.error.removeObservers(lifecycleOwner)
    }
}
